calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,367 @@
+use crate::analysis::ReadWriteSet;
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+use crate::ir::{self, CloneName};
+use std::collections::BTreeMap;
+
+#[derive(Default)]
+/// Transforms a group into a seq of 2 smaller groups, if possible.
+/// Currently, in order for a group to be transformed must
+/// 1) Group must write to exactly 2 cells -- let's call them cell1 and cell2
+/// 2) cell1 and cell2 must be either non-combinational primitives or components
+/// 3) Must have group[done] = cell2.done and cell2.go = cell1.done;
+/// 4) All reads of cell1 must be a stable port or cell1.done.
+pub struct GroupToSeq {
+    ///Maps names of group to the sequences that will replace them
+    group_seq_map: BTreeMap<ir::Id, ir::Control>,
+}
+
+impl Named for GroupToSeq {
+    fn name() -> &'static str {
+        "group2seq"
+    }
+
+    fn description() -> &'static str {
+        "split groups under correct conditions"
+    }
+}
+
+impl Visitor for GroupToSeq {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        for g in groups.iter() {
+            let mut group = g.borrow_mut();
+            match SplitAnalysis::get_split(
```

Comment:
It seems like `SplitAnalysis` returns two groups in the `Ok` case. What if we make it take the `group` directly instead of passing the assignments and the name? That way, we also avoid this `drain(..).collect()` call.